backend/apps/rag/batch_utils.py

```diff
@@ -0,0 +1,258 @@
+            current_tokens += tokens
+    if current_request:
+        yield (
+            batch_idx,
+            b"\n".join(
+                _request_iterator(current_request, model, uuid, batch_idx, chunk_size)
+            ),
+        )
+
+
+@retry(
+    stop=stop_after_attempt(TENAICITY_RETRY_ATTEMPTS),
+    wait=wait_exponential(
+        multiplier=TENAICITY_WAIT_MULTIPLIER,
+        min=TENAICITY_WAIT_MIN,
+        max=TENAICITY_WAIT_MAX,
+    ),
+)
+def _process_one_batch(
+    client: OpenAI,
+    uuid: str,
+    batch_idx: int,
+    batch: bytes,
+    completion_window: Literal["24h"],
+    wait_interval: int,
+) -> list[list[float]]:
+    # Process a single batch of requests
+    # This function creates a batch job, waits for it to complete, and returns the embeddings
+    # The embeddings are sorted by the custom_id, which includes the batch index and request index
+    # This ensures that the embeddings are returned in the same order as the input texts
+    try:
+        input_file: FileObject = client.files.create(
+            file=(f"batch_{uuid}_{batch_idx}.jsonl", batch), purpose="batch"
+        )
+        batch_job: Batch = client.batches.create(
+            input_file_id=input_file.id,
+            endpoint="/v1/embeddings",
+            completion_window=completion_window,
+        )
+        while True:
+            batch_status: Batch = client.batches.retrieve(batch_job.id)
+            status: Literal[
+                "validating",
+                "failed",
+                "in_progress",
+                "finalizing",
+                "completed",
+                "expired",
+                "cancelling",
+                "cancelled",
+            ] = batch_status.status
+            output_file_id = batch_status.output_file_id
+
+            if output_file_id and status == "completed":
+                log.info("Batch Job Completed")
+                output_content: str = client.files.content(output_file_id).text
+                raw_results = map(loads, output_content.strip().split("\n"))
+                sorted_results = sorted(raw_results, key=itemgetter("custom_id"))
+                return [
+                    data["embedding"]
+                    for result in sorted_results
+                    for data in result["response"]["body"]["data"]
+                ]
+
+            elif status in ("failed", "expired", "cancelled"):
+                raise RuntimeError(f"Batch Job Failed: {status}")
+
+            _print_with_time(f"Batch Request Status: {status}")
+            sleep(wait_interval)
+    except Exception as e:
+        log.exception(e)
+        raise
+
+
+def _request_async_embedding(
+    model: str,
+    texts: list[str],
+    key: str,
+    url: str,
+    completion_window: Literal["24h"],
+    wait_interval: int,
+    token_limit: int,
+) -> list[list[float]]:
+    uuid: str = uuid4().hex
+    client: OpenAI = OpenAI(api_key=key, base_url=url)
+    return [
```

Comment:
(embedding
        for batch_idx, batch in _batch_iterator(texts, model, uuid, token_limit)
        for embedding in _process_one_batch(
            client, uuid, batch_idx, batch, completion_window, wait_interval
        )
    ) it's better ti use Generator instead of list